### PR TITLE
Provides functionality to update a model at execution time

### PR DESCRIPTION
**Pull request by koparasy**

### CODE REVIEW

File: src/AMSlib/ml/surrogate.hpp [hunk 1]
```diff
@@ -393,6 +393,23 @@ class SurrogateModel
...
 
   bool is_DeltaUQ() { return _is_DeltaUQ; }
+
+  void update(std::string new_path)
+  {
+    /* This function updates the underlying torch model,
+     * with a new one pointed at location modelPath. The previous
+     * one is destructed automatically.
+     *
+     * TODO: I decided to not update the model path on the ``instances''
```
Comment (lpottier) [inline src/AMSlib/ml/surrogate.hpp:403]: I think that's okay for now. However, we should definitely track internally at least for our own tests.

File: src/AMSlib/wf/basedb.hpp [hunk 2]
```diff
@@ -112,6 +112,8 @@ class BaseDB
                      std::vector<TypeValue*>& outputs) = 0;
 
   uint64_t getId() const { return id; }
+
+  virtual bool updateModel() { return false; }
```
Comment (lpottier) [inline src/AMSlib/wf/basedb.hpp:116]: I am not sure managing that with a `bool` is enough. With this model, the first time we update it's easy, I set the `bool` to `true` but then for the second updates I have to set it to false once the first update went through.
We might have to store the latest model path used for example, so we know that we have to update the model if paths are different.

Note that maybe it's for me to implement that part and you just picked a simple way of testing the model update, if that's the case disregard that comment :)

**Review by lpottier — APPROVED** [review]
Th PR looks good to me!

File: src/AMSlib/wf/basedb.hpp (hunk 2 above)
Comment (koparasy) [reply src/AMSlib/wf/basedb.hpp:116]: I agree, I thought we should re-iterate this. But we need to return something. 

Maybe we should do something among the lines of:
`model.getCurrentModel() != DB.mostRecentModel()`

We should reiterate this part once we have the mechanism in place to receive a new model.

File: src/AMSlib/ml/surrogate.hpp [hunk 3]
```diff
@@ -393,6 +393,23 @@ class SurrogateModel
   }
 
   bool is_DeltaUQ() { return _is_DeltaUQ; }
+
+  void update(std::string new_path)
```
Comment (ggeorgakoudis) [inline src/AMSlib/ml/surrogate.hpp:397]: reference?

File: src/AMSlib/ml/surrogate.hpp (hunk 1 above)
Comment (koparasy) [reply src/AMSlib/ml/surrogate.hpp:403]: I agree, I wasn't sure if this needs to be tracked by the DB, the WF or some other entity that tracks the provenance.

File: src/AMSlib/ml/uq.hpp [hunk 4]
```diff
@@ -142,6 +140,23 @@ class UQ
     }
   }
 
+  void updateModel(std::string model_path, std::string uq_path = "")
```
Comment (ggeorgakoudis) [inline src/AMSlib/ml/uq.hpp:143]: uq_path is unused?

Comment (koparasy) [reply src/AMSlib/ml/uq.hpp:143]: For FAISS we need 2 paths. Cause those are encoded by separate entities. For DeltaUQ the model includes UQ.

File: src/AMSlib/ml/surrogate.hpp (hunk 3 above)
Comment (koparasy) [reply src/AMSlib/ml/surrogate.hpp:397]: You are right. I Will fix this on a later iteration. Thought you will not review, and @lpottier  did not catch this and I merged once he landed his review.

Comment (lpottier) [reply src/AMSlib/ml/surrogate.hpp:397]: Missed this one indeed (albeit the performance difference should be negligible)

**Pull request merged**